src/components/tl/cuda/tl_cuda_topo.c

```diff
@@ -5,172 +5,293 @@
+    cuda_st = cudaGetDeviceCount(&num_gpus);
+    if ((cuda_st != cudaSuccess) || (num_gpus == 0)){
+        tl_info(topo->lib, "cudaGetDeviceCount failed or no GPU devices found");
+        return UCC_ERR_NO_RESOURCE;
+    }
+    topo->num_nodes = 0;
+    topo->graph = (ucc_tl_cuda_topo_node_t*)ucc_malloc(MAX_PCI_DEVICES *
+                                                       sizeof(*topo->graph),
+                                                       "tl cuda topo graph");
+    if (!topo->graph) {
+        tl_error(topo->lib, "failed to allocate tl cuda topo graph");
+        return UCC_ERR_NO_MEMORY;
+    }
+    kh_init_inplace(bus_to_node, &topo->bus_to_node_hash);
+    pthread_mutex_lock(&nvml_lock);
+    NVMLCHECK_GOTO(nvmlInit_v2(), exit_free_graph, status, topo->lib);
+    nvml_value.fieldId = NVML_FI_DEV_NVLINK_LINK_COUNT;
+    for (i = 0; i < num_gpus; i++) {
+        CUDACHECK_GOTO(cudaDeviceGetPCIBusId(pci_bus_str, MAX_PCI_BUS_ID_STR, i),
+                       exit_nvml_shutdown, status, topo->lib);
+        status = ucc_tl_cuda_topo_get_pci_id(topo->lib, i, &pci_id);
+        if (status != UCC_OK) {
+            goto exit_nvml_shutdown;
+        }
+        status = ucc_tl_cuda_topo_graph_add_device(&pci_id, topo, &node);
+        if (status != UCC_OK) {
+            goto exit_nvml_shutdown;
+        }
+        NVMLCHECK_GOTO(nvmlDeviceGetHandleByPciBusId_v2(pci_bus_str, &nvml_dev),
+                       exit_nvml_shutdown, status, topo->lib);
+        NVMLCHECK_GOTO(nvmlDeviceGetFieldValues(nvml_dev, 1, &nvml_value),
+                       exit_nvml_shutdown, status, topo->lib);
+        num_nvlinks = ((nvml_value.nvmlReturn == NVML_SUCCESS) &&
+                       (nvml_value.valueType == NVML_VALUE_TYPE_UNSIGNED_INT)) ?
+                      nvml_value.value.uiVal : 0;
+        for (link = 0; link < num_nvlinks; link++) {
+            NVMLCHECK_GOTO(nvmlDeviceGetNvLinkRemoteDeviceType(nvml_dev, link,
+                                                               &nvml_dev_type),
+                           exit_nvml_shutdown, status, topo->lib);
+            if ((nvml_dev_type != NVML_NVLINK_DEVICE_TYPE_GPU) &&
+                (nvml_dev_type != NVML_NVLINK_DEVICE_TYPE_SWITCH))
+            {
+                /* nvlink connected device is not supported by cuda tl */
                 continue;
             }
-            for (k = 0; k < size; k++) {
-                if (ucc_tl_cuda_topo_is_direct(team, topo, i, k) &&
-                    ucc_tl_cuda_topo_is_direct(team, topo, k, j)) {
-                    topo->proxies[p].src   = i;
-                    topo->proxies[p].dst   = j;
-                    topo->proxies[p].proxy = k;
-                    break;
-                }
+            NVMLCHECK_GOTO(nvmlDeviceGetNvLinkRemotePciInfo_v2(nvml_dev, link,
+                                                               &nvml_pci),
+                           exit_nvml_shutdown, status, topo->lib);
```

Comment:
what it will return if remote is NVML_NVLINK_DEVICE_TYPE_SWITCH ?

Reply:
it will return pcie address of the switch